app/src/main/java/com/codesoom/assignment/service/ProductService.java

```diff
@@ -0,0 +1,61 @@
+package com.codesoom.assignment.service;
+
+import com.codesoom.assignment.domain.Product;
+import com.codesoom.assignment.dto.ProductResponse;
+import com.codesoom.assignment.exception.ProductNotFoundException;
+import com.codesoom.assignment.repository.ProductRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+@Transactional(readOnly = true)
```

Comment:
이 readOnly는 어떤 일을 하는걸까요?

readOnly 안에는 어떤 마법이 일어나는 걸까요? 

한번 알아보면 좋을 것 같아요!

또한 현재 테스트에서 유효한지도요!

Reply:
- 영속성 컨텍스트는 변경을 감지하는 스냅샷을 저장하지만 `readOnly = true`로 설정하면 저장하지 않아 메모리를 아끼는 성능 향상을 꾀할 수 있다.
- 하지만 동시에 `flush`를 `Manual`로 바꾸어서 read, write가 동작하지 않는다.

위 코드에서 create, delete가 있으므로 `readOnly = true`를 지우는 게 맞는 것일까요~?

Reply:
좋아요! 하지만 이 내용의 출처를 정확히 기술해보는 것이 좋습니다! 그리고 출처는 블로그보다는 도큐먼트가 선호됩니다!

또한 readOnly = true를 지우기보다 true인 것과 아닌 것을 분리하는 것이 좋겠어요!

Reply:
그리고 우리가 테스트로 사용하는 H2에는 readOnly=true를 잘 지원하지 않습니다 ㅠㅜㅠㅜ 
이것도 왜 그런지 알아보면 재밌을 거에요!

Reply:
오호 넵넵 이 경우에 [도큐먼트](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Transactional.html)를 참조헀는데 런타임에 응답을 최적화한다는 것으로 이해했습니다.
그런데 최적화하기 위해 뭘 하는지는 이해할 수 없어서 다른 블로그들을 참조했는데 아마 [자바 ORM 표준 JPA 프로그래밍](https://www.aladin.co.kr/shop/wproduct.aspx?ItemId=62681446) 책에서 위 내용이 나와서 몇 다리를 걸쳐서 제가 알게 된 것 같습니다.

위 사례를 일반화하여서..
도큐먼트를 참조해야 한다는 조언이 실제 사례에서 저와 같은 초보자에게도 통하는 게 맞는지 궁금합니다!
구체적인 적용의 경험이 있다면 도큐먼트를 봐도 이해가 될 텐데 그게 아니라면 구체적인 적용을 해본 사람이 쓴 책이나 블로그를 봐야 이해가 되는 경우가 많았던 것 같습니다.
그래서 경력이 쌓일수록 도큐먼트로 참조를 옮겨가는 게 더 맞지 않아 싶어서 질문을 남겨 봅니다~

H2에서도 도큐먼트에서는 `readOny = true`를 왜 잘 지원하지 않는지는 찾을 수가 없었습니다.. ㅠ.ㅠ~

Reply:
맞아요! 멋진 책과 블로그가 우리에게 많은 도움을 주죠. 그리고 정리된 내용들을 읽으면 우리에게 무언가를 이해할 때 빠르게 해결 할 수 있어요! 
여기서 저는 블로그나 책을 보지말고 도큐먼트만 보세요! 라고 말할 수는 없어요!

하지만 당연히 아시겠지만, 모든 기능은 시간이 지남에 따라 개선되고 변경되고 진화하지만 블로그와 책은 발간된 시간에서 멈춰있다는 것을요 ㅠㅜㅠㅜ

아마 연차가 쌓이면서 자연스럽게 도큐먼트를 옮겨간다는 것은 연차가 쌓여서 '이제 도큐먼트를 읽을 만큼의 실력이 쌓였어!' 가 아니라 '결국 도큐먼트, 공식문서가 가장 유효한 글이구나...' 라는 경험이 쌓이면서 도큐먼트를 읽게되는 것 같아요 ㅠㅜㅠㅜ

빠르게 더 먼저 빠르게 읽어보는 것이 좋습니다. 하지만 도큐먼트가 별로인 프로젝트도 있어요. 그런 곳은 오히려 블로그가 도움이 되기도합니다.

Reply:
읽어보신 도큐먼트를 잠깐보자면 readOnly 정보에 아래와 같은 글이 있습니다.
`This just serves as a hint for the actual transaction subsystem; it will not necessarily cause failure of write access attempts. A transaction manager which cannot interpret the read-only hint will not throw an exception when asked for a read-only transaction but rather silently ignore the hint.`

Reply:
여기서 우리가 알아야할 점은 그저 힌트만 준다는 것입니다. H2 구현체는 이 힌트를 무시하고 있습니다.

Reply:
```java
  /**
    * According to the JDBC specs, this setting is only a hint to the database
    * to enable optimizations - it does not cause writes to be prohibited.
    *
    * @param readOnly ignored
    * @throws SQLException if the connection is closed
    */
  @Override
  public void setReadOnly(boolean readOnly) throws SQLException {
      try {
	  if (isDebugEnabled()) {
	      debugCode("setReadOnly(" + readOnly + ");");
	  }
	  checkClosed();
      } catch (Exception e) {
	  throw logAndConvert(e);
      }
  }
```

Reply:
H2 구현체 주석에도 나오듯이 
JDBC 스펙에 따르면, readOnly는 그저 최적화를 위한 힌트일 뿐이다. 또한 readOnly는 쓰기가 금지되는 것되는 것도 아니다. 

주석아래 `@param readOnly ignored` 를 보면 readOnly는 무시된다는 것을 알 수 있습니다.

Reply:
오후 그렇군요. 
원칙은 도큐먼트, 나머지 자료는 취사선택, 핵심은 빠르게. 
원칙에 매몰되어서는 안 되지만 그래도 기본은 원칙이다~

[h2 문서](http://www.h2database.com/html/main.html)가 아니라 코드 주석을 보고 알 수 있었군요..! 
어우 가르침 감사드립니다!